fix: DiffRemote: close_diffview_mapping command updated

alexrah opened the pull request:
call diffoff also for the source buffer to avoid keeping diff-mode open

alexrah committed:
fix: DiffRemote: close_diffview_mapping command updated
call diffoff also for the source buffer to avoid keeping diff-mode open

alexrah commented:
fixes #27

coffebar commented:
Hi! I want to suggest an alternative approach that might be more reliable.

### Potential fix consideration:
  Instead of command chaining like <cmd>diffoff | bd! | diffoff<cr> (which has
  unpredictable window focus behavior after bd!), tracking both window IDs before
  opening the diff ensures reliable cleanup:

```lua
  local orig_win = vim.api.nvim_get_current_win()
  vim.api.nvim_command("silent! diffsplit " .. remote_path)
  local remote_win = vim.api.nvim_get_current_win()

  -- In the close mapping:
  vim.keymap.set("n", config.options.close_diffview_mapping, function()
    if vim.api.nvim_win_is_valid(orig_win) then
      vim.api.nvim_win_call(orig_win, function() vim.cmd("diffoff") end)
    end
    if vim.api.nvim_win_is_valid(remote_win) then
      vim.api.nvim_win_call(remote_win, function() vim.cmd("diffoff") end)
    end
    vim.cmd("bd!")
  end, { buffer = true })
```
  I've created a patch with the full implementation if you'd like to review:

[diffremote-fix.patch](https://github.com/user-attachments/files/23171102/diffremote-fix.patch)

  This patch replaces the simple command chain approach with explicit window
  tracking to ensure both diff windows exit diff mode cleanly.

  Why tracked windows instead of <cmd>diffoff | bd! | diffoff<cr>:
  - After bd! closes the buffer/window, the second diffoff may not execute in the
  correct window due to unpredictable focus switching
  - The command chain behavior varies depending on window layout and Vim settings
  - Tracking window IDs with nvim_win_call() guarantees both specific windows are
  targeted, regardless of focus changes
  - Includes validity checks to handle manually closed windows gracefully

Status: Conceptual improvement, not tested yet.

To apply patch:
```
cd /path/to/transfer.nvim
git apply /tmp/diffremote-fix.patch
```

Currently, I don't use this plugin and don't have envs ready to test on. My patch is generated with a help of AI. 
Please check it out, I hope it will work. If so, you can consider update PR with it.

alexrah commented:
yeah, that's a sound idea. I'm going to test it and if it works, update the PR

alexrah committed:
fix: DiffRemote: close_diffview_mapping command reworked
replace diffoff command chain with single diffoff calls targeting the actual windows

coffebar commented:
Is it ready to merge?

alexrah commented:
I had to rework it a little. I've also simplified the logic:
- tracking `remote_win` in this context is not necessary
- this disposable autocmd already run only into the buffer opened by `DiffRemote`, which is by design the remote buffer

after this rework the original `diffoff` & `bd!` keep the same logic as before, while the added `diffoff` tracks the window where the `DiffRemote` command was executed, avoiding to accidentally close the wrong window

alexrah commented:
> Is it ready to merge?

Well, I've tested on my setup and now it's working without any issues
